### ISSUE TITLE
[Enhancement] Embedded HTML code parsing

### ISSUE THREAD

**Issue by donmor** [Zettlr/Zettlr#2902]
### Description

We all know that Markdowns are based on HTML. Sometimes there're some files including embedded HTML codes like this:
![2021-12-08 22-35-35屏幕截图](https://user-images.githubusercontent.com/29735136/145226898-5d86ebac-da1c-47a1-b7a3-49e35318d94b.png)
The file is from one of my projects on GitHub. These HTML codes and escapes should be rendered as HTML content, but Zettlr rendered them as source code.

### Proposed Changes

These HTML codes should be properly rendered, just like how they're rendered by GitHub.

### Caveats

For people that satisfied with current style, there should be an item in preferences to allow render embedded HTML as source code.

### Do you Wish to Attempt Implementing this Yourself?

Yes

### Zettlr Version

Stable (most recent version)

### Your Platform

- [ ] Windows
- [ ] macOS
- [X] Linux

### Operating System Version

Ubuntu 20.04 LTS

**Comment by nathanlesage** [comment]
Ooooh, very slippery slope here. I've gotten a CVE earlier this year because I allowed YouTube embeds and people figured out you could hijack computers using this. Cost me a few gray hairs and a few weeks to fix. I'm definitely NOT in favor of this at this moment.

**Comment by donmor** [comment]
Not all HTML should be parsed. Just whitelist the tags considered not "active" like `<span>` and `<table>`, and escapes like `&lt;` for `<`. Also, remove the tags that are not whitelisted and their inner text, for example, `<script>some script</script>`will entirely invisible.

**Comment by nathanlesage** [comment]
> Just whitelist the tags considered not "active" like <span> and <table>, and escapes like &lt; for <

That's not how it works. You can still attach event listeners executing harmful content to these tags. Believe me, there would already be much bigger support for embeds if it were that easy.

**Comment by donmor** [comment]
> > Just whitelist the tags considered not "active" like `<span>` and `<table>`, and escapes like < for `<`
> 
> That's not how it works. You can still attach event listeners executing harmful content to these tags. Believe me, there would already be much bigger support for embeds if it were that easy.

So what about making another parser that only parses "safe" tags? Inventing wheels though.

**Comment by donmor** [comment]
Or a preprocessor that removes unsafe tags?

**Comment by donmor** [comment]
An example is TiddlyWiki, which disables active comtents in regular tiddlers.

**Comment by nathanlesage** [comment]
The thing is, HTML does not have a concept of "active" or "passive" components, so that won't help. I mean, it is definitely possible to render HTML elements, but it requires A LOT of effort. The reason is that there is a potentially infinite amount of ways malicious actors can hide bad code in there, and we are then basically left with three options:

1. Don't render HTML. That is what we do right now and it is both easy and totally safe.
2. Render only "safe" HTML. However, due to the magnitude of potential attack vectors, we would basically end up coding an insane amount of rules in there – strip all attributes except the ones necessary (which ones are necessary?), remove script tags (possibly the easiest to accomplish), and the likes
3. Try to figure out what the corresponding code wants to do. Does it want to just add an image? Why not transform it to Markdown? Etc.

As you see, options 2 and 3 are insanely complex and would require a security-related knowledge which I simply don't possess. I also tried to go the way of "Hey, just be careful with what you do", but the attack vectors laid out in the earlier CVE basically relied on a simple fact I had overlooked: Just send someone a malicious Markdown file and have them open it in Zettlr.

So I still don't see a way of doing this safely.
